File: modules/messaging/src/main/java/org/torquebox/messaging/MessageProcessor.java
/*
 * Copyright 2008-2013 Red Hat, Inc, and individual contributors.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package org.torquebox.messaging;

import javax.jms.Message;

import org.hornetq.api.core.client.ClientMessage;
import org.jboss.logging.Logger;
import org.jruby.Ruby;
import org.jruby.runtime.builtin.IRubyObject;
import org.projectodd.polyglot.messaging.BaseMessageProcessor;
import org.torquebox.core.util.RuntimeHelper;
import org.torquebox.messaging.component.MessageProcessorComponent;

public class MessageProcessor extends BaseMessageProcessor {

    @Override
    public void onMessage(Message message) {
        if (getConsumer() == null) {
            return; // racist!
        }
        MessageProcessorGroup group = (MessageProcessorGroup) getGroup();
        try {
            ensureCurrentRuntime();
            if (getConsumer() == null) {
                return; // racist!
            }

            MessageProcessorComponent component = (MessageProcessorComponent) group.getComponentResolver().resolve( this.currentRuby );
            try {
                component.process( message, getSession(), group );
                if (isXAEnabled()) {
                    log.trace( "Committing XA transaction for messageId " + message.getJMSMessageID() );
                    commitXATransaction();
                } else {
                    log.trace( "Committing non-XA transaction for messageId " + message.getJMSMessageID() );
                    getSession().commit();
                }
            } catch (Throwable e) {
                if (isXAEnabled()) {
                    log.trace( "Rolling back XA transaction for messageId " + message.getJMSMessageID() );
                    rollbackXATransaction( e );
                } else {
                    log.trace( "Rolling back non-XA transaction for messageId " + message.getJMSMessageID() );
                    getSession().rollback();
                    throw e;
                }
            }
        } catch (Throwable e) {
            log.error( "Unexpected error in " + group.getName(), e );
        } finally {
            clearCurrentRuntime();
        }
    }

    @Override
    protected void prepareTransaction() {
        try {
            ensureCurrentRuntime();
            RuntimeHelper.call( this.currentRuby, this.currentRubyTM, "prepare", new Object[] {} );
            RuntimeHelper.call( this.currentRuby, this.currentRubyTM, "enlist", new Object[] { getSession() } );
        } catch (Exception e) {
            log.error( "Failed to prepare transaction for message", e );
        }
    }

    protected void rollbackXATransaction(Throwable e) {
        if (this.currentRuby != null && this.currentRubyTM != null) {
            RuntimeHelper.call( this.currentRuby, this.currentRubyTM, "error", new Object[] { e } );
        }
    }

    protected void commitXATransaction() {
        if (this.currentRuby != null && this.currentRubyTM != null) {
            RuntimeHelper.call( this.currentRuby, this.currentRubyTM, "commit", new Object[] {} );
        }
    }

    protected void ensureCurrentRuntime() throws Exception {
        if (this.currentRuby == null) {
            this.currentRuby = ((MessageProcessorGroup) getGroup()).getRubyRuntimePool().borrowRuntime( getGroup().getName() );
            this.currentRubyTM = RuntimeHelper.invokeClassMethod( this.currentRuby, "TorqueBox::Transactions::Manager", "current", new Object[] {} );
        }
    }

    protected void clearCurrentRuntime() {
        if (this.currentRuby != null) {
            try {
                RuntimeHelper.evalScriptlet( this.currentRuby, "Thread.current[:torquebox_transaction] = nil" );
                ((MessageProcessorGroup) getGroup()).getRubyRuntimePool().returnRuntime( this.currentRuby );
            } catch (Throwable e) {
                log.warn( "Possible memory leak?", e );
            }
        }
        this.currentRuby = null;
        this.currentRubyTM = null;
    }

    public static final Logger log = Logger.getLogger( "org.torquebox.messaging" );

    private Ruby currentRuby = null;
    private Object currentRubyTM = null;
}

File: gems/messaging/lib/torquebox/messaging/message_processor.rb
# Copyright 2008-2013 Red Hat, Inc, and individual contributors.
# 
# This is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation; either version 2.1 of
# the License, or (at your option) any later version.
# 
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this software; if not, write to the Free
# Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA, or see the FSF site: http://www.fsf.org.

require 'torquebox/messaging/processor_middleware/default_middleware'

module TorqueBox
  module Messaging
    class MessageProcessor
      include ProcessorMiddleware::DefaultMiddleware
      
      attr_accessor :message

      def initialize
        @message = nil
        @proxy = nil
      end

      def initialize_proxy(group)
        @proxy = MessageProcessorProxy.new(group)
      end

      def method_missing(method, *args, &block)
        @proxy.send( method, *args, &block )
      end

      def on_message(body)
        throw "Your subclass must implement on_message(body)"
      end

      def on_error(error)
        raise error
      end

      def process!(message)
        @message = message
        begin
          value = on_message(message.decode)
          reply(value) if synchronous?
        rescue Exception => e
          on_error( e ) 
        end 
      end

      def reply(value)
        TorqueBox::Messaging::Queue.new(@message.jms_message.jms_destination.queue_name).publish(value, :correlation_id => @message.jms_message.jms_message_id)
      end

      class << self

        # List all available message processors for current application.
        #
        # @return [Array<TorqueBox::Messaging::MessageProcessorProxy>] List of
        #         proxy objets to read and manage state of selected message
        #         processor
        def list
          processors = []

          TorqueBox::MSC.get_services(/^#{messaging_service_name.canonical_name}\.\".*\"$/) do |service|
            processors << MessageProcessorProxy.new(service.value)
          end

          processors
        end

        # Lookup a message processor by its destination and class name.
        #
        # @param [String] The destination name (queue, topic) to which
        #         a message processor is bound.
        #
        # @param [String] The class name of the message processor
        #         implementation.
        def lookup(destination_name, class_name)
          sn = messaging_service_name.append("#{destination_name}.#{class_name}")

          # Try to find a message procesor for specified parameters
          group = TorqueBox::ServiceRegistry::lookup(sn)

          return MessageProcessorProxy.new(group) if group

          # Ooops, no processor is found. Most probably wrong data.
          return nil
        end

        protected

        def messaging_service_name
          TorqueBox::MSC.deployment_unit.service_name.append('torquebox').append('messaging')
        end
      end
    end

    class MessageProcessorProxy
      def initialize(group)
        @group = group

        raise "Cannot create MessageProcessorProxy for non-existing MessageProcessorGroup" if @group.nil?
      end

      attr_reader :destination_name, :class_name

      # Updates the concurrency,
      #
      # @note This method sets the concurrency and changes immediately
      #       the number of consumers for specified destination.
      def concurrency=(size)
        raise "Setting concurrency for '#{name}' to value < 0 is not allowed. You tried '#{size}'." if size < 0

        return size if size == @group.concurrency

        @group.update_concurrency(size)

        concurrency
      end

      # Returns the concurrency
      #
      # @return Integer
      def concurrency
        @group.concurrency
      end

      # Returns the group name
      #
      # @return String
      def name
        @group.name
      end

      # Returns the destination (queue or topic) name
      #
      # @return String
      def destination_name
        @group.destination_name
      end

      # Returns the message processor implementation
      # class name
      #
      # @return String
      def class_name
        @group.message_processor_class.name
      end

      # Returns the message selector
      #
      # If there is no message selector specified,
      # returns empty string
      #
      # @return String
      def message_selector
        @group.message_selector
      end

      # Returns true if the message processor is a durable
      # subscriber, false otherwise
      #
      # @return Boolean
      def durable?
        @group.durable
      end

      # Returns true if the message processor is synchronous,
      # false otherwise
      #
      # @return Boolean
      def synchronous?
        @group.synchronous
      end

      def to_s
        "[MessageProcessorProxy: #{name}]"
      end
    end
  end
end
